Added review workflow

itiannerona opened the pull request:
Possible workflow for #4.

itiannerona committed:
Replaced ClassData with TheoryData

itiannerona committed:
Updated dev with main

itiannerona committed:
Updated Account to pass test

itiannerona committed:
Reverted comparison change on Journal Entry and corrected test data for AccountsAreExclusiveToDebitOrCreditSide

itiannerona committed:
Added required member UserId to JournalEntry and to tests

itiannerona committed:
Added review model and condition on adding reviews to journal entry

maratmuginov commented on source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs:
```diff
@@ -35,15 +39,37 @@ public record JournalEntry
+            if (Closed)
+                throw new InvalidOperationException("Journal is already approved and closed");
+
+            if (string.Compare(UserId, review.Reviewer, true) == 0)
```
Strings use value equality semantics, why are we using .Compare instead of `==` or .Equals?

maratmuginov commented on source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs:
```diff
@@ -35,15 +39,37 @@ public record JournalEntry
+                throw new ArgumentException("Accountant cannot review and approve own journal entry", nameof(review.Reviewer));
+
+            if (review.Approved)
+                Closed = true; //maybe emit event so a service can save journal
```
What component would benefit from this event?

itiannerona replied:
It is just a suggestion. This is about a pattern I read sometime ago, but we do not have to implement it if you don't see it befitting. The idea is to make services subscribe to an event that is triggered whenever a journal entry is approved. it could be a service that persists a journal entry to the db, and another service that sends an email or a notification on the screen (perhaps through SignalR) of a user, instead of calling or instantiating those services. My understanding of events and design patterns is shallow, so I wanted to see what you thought about it.

maratmuginov replied:
Ah I like this! I think it's a good idea to use events, but let's wait until we have someone who might consume them. But otherwise, events are pretty simple in dotnet thankfully. I appreciate the suggestion.

maratmuginov commented on source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs:
```diff
@@ -35,15 +39,37 @@ public record JournalEntry
+            if (review.Approved)
+                Closed = true; //maybe emit event so a service can save journal
+
+            if (Reviews is null)
```
Why not just create an empty list at initialization?

maratmuginov commented on source/Pacioli/Pacioli.Lib/Models/Review.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        public Review(string userId, bool approved, string comment) : this(userId, approved)
+        {
+            if (string.IsNullOrEmpty(comment))
```
I think we shouldn't require comments, just userId and approval should be enough.

maratmuginov reviewed (changes requested):
Looks good, just a small change in equality comparison and I don't think we should require comments in reviews.

itiannerona committed:
Addressing previous comments.

itiannerona committed:
Removed Comment not empty constraint on Review

itiannerona commented:
If you are wondering where the .Net build failed messages came from, it showed when I added an action, and when I pushed something, it threw an error on ``dotnet restore`` saying ``MSBUILD : error MSB1003: Specify a project or solution file.
The current working directory does not contain a project or solution file.``. I deleted the action for now.

I added the action as a direct push to main, probably the action runner does not like that. Also, I added an action because on the this PR, I was able to push a failing test (immutability). Btw the immutability test is another thing that I want to ask about.
